overlord/devicestate/remodel_test.go

```diff
@@ -0,0 +1,704 @@
+		defer s.state.Unlock()
+		assertstate.ReplaceDB(s.state, db)
+
+		assertstatetest.AddMany(s.state, s.storeSigning.StoreAccountKey(""))
+		assertstatetest.AddMany(s.state, s.brands.AccountsAndKeys("my-brand")...)
+	}()
+
+	s.dummyStore = new(storetest.Store)
+
+	newStore := func(devBE storecontext.DeviceBackend) snapstate.StoreService {
+		s.capturedDevBE = devBE
+		return s.dummyStore
+	}
+
+	hookMgr, err := hookstate.Manager(s.state, o.TaskRunner())
+	c.Assert(err, IsNil)
+	s.mgr, err = devicestate.Manager(s.state, hookMgr, o.TaskRunner(), newStore)
+	c.Assert(err, IsNil)
+}
+
+var modelDefaults = map[string]interface{}{
+	"architecture":   "amd64",
+	"kernel":         "my-brand-kernel",
+	"gadget":         "my-brand-gadget",
+	"store":          "my-brand-store",
+	"required-snaps": []interface{}{"required1"},
+}
+
+func fakeRemodelingModel(extra map[string]interface{}) *asserts.Model {
+	primary := map[string]interface{}{
+		"type":         "model",
+		"authority-id": "my-brand",
+		"series":       "16",
+		"brand-id":     "my-brand",
+		"model":        "my-model",
+	}
+	return assertstest.FakeAssertion(primary, modelDefaults, extra).(*asserts.Model)
+}
+
+func (s *remodelLogicSuite) TestClassifyRemodel(c *C) {
+	oldModel := fakeRemodelingModel(nil)
+
+	cases := []struct {
+		newHeaders map[string]interface{}
+		kind       devicestate.RemodelKind
+	}{
+		{map[string]interface{}{}, devicestate.UpdateRemodel},
+		{map[string]interface{}{
+			"required-snaps": []interface{}{"required1", "required2"},
+			"revision":       "1",
+		}, devicestate.UpdateRemodel},
+		{map[string]interface{}{
+			"store":    "my-other-store",
+			"revision": "1",
+		}, devicestate.StoreSwitchRemodel},
+		{map[string]interface{}{
+			"model": "my-other-model",
+			"store": "my-other-store",
+		}, devicestate.ReregRemodel},
+		{map[string]interface{}{
+			"authority-id": "other-brand",
+			"brand-id":     "other-brand",
+			"model":        "other-model",
+		}, devicestate.ReregRemodel},
+		{map[string]interface{}{
+			"authority-id":   "other-brand",
+			"brand-id":       "other-brand",
+			"model":          "other-model",
+			"required-snaps": []interface{}{"other-required1"},
+		}, devicestate.ReregRemodel},
+		{map[string]interface{}{
+			"authority-id": "other-brand",
+			"brand-id":     "other-brand",
+			"model":        "other-model",
+			"store":        "my-other-store",
+		}, devicestate.ReregRemodel},
+	}
+
+	for _, t := range cases {
+		newModel := fakeRemodelingModel(t.newHeaders)
+		c.Check(devicestate.ClassifyRemodel(oldModel, newModel), Equals, t.kind)
+	}
+}
+
+func (s *remodelLogicSuite) TestUpdateRemodelContext(c *C) {
+	oldModel := fakeRemodelingModel(nil)
+	newModel := fakeRemodelingModel(map[string]interface{}{
+		"required-snaps": []interface{}{"required1", "required2"},
+		"revision":       "1",
+	})
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	remodCtx, err := devicestate.RemodelCtx(s.state, oldModel, newModel)
+	c.Assert(err, IsNil)
+
```

Comment:
Would it make sense to explicitly check for the type here? Its unexport so as a quick idea something like:
```go
	c.Check(fmt.Sprintf("%T", remodCtx), Equals, "devicestate.updateRemodelContext")
```
or too hacky/crazy?

Reply:
the follow up adds a Kind method to remodelContext and does this:

c.Check(remodCtx.Kind(), Equals, devicestate.UpdateRemodel)